jedi/evaluate/pep0484.py

```diff
@@ -82,7 +82,58 @@ def _fix_forward_reference(context, node):
 
 @memoize_default()
 def infer_param(execution_context, param):
```

Comment:
Hmm, I think you're doing too much work here for no gain. I think this should be done with the parser. The parser can do a lot of cool stuff. By parsing something like `parse('foo(%s): pass' % regex_string)`, you could really easy validate a lot of things that you assume here or test with complicated param matching.